ossf/scorecard
# :sparkles: Add machine-readable patch to fix script injections in workflows

#### What kind of change does this PR introduce?

- [x] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?

Findings have a `.Remediation.Patch` field which is meant to contain a machine-readable patch fixing that particular finding:

https://github.com/ossf/scorecard/blob/3155309aa81adf3395f4d62ee133b524ff316da1/finding/probe.go#L50-L52

This field currently isn't used by any Scorecard probe.

#### What is the new behavior (if this is a feature change)?**

This PR adds a machine-readable patch (following the "unified diff" format which users can then apply using `patch` or `git apply`) fixing all `hasDangerousWorkflowScriptInjection` findings.

Each finding is patched by creating a global environment variable that wraps the dangerous GitHub variable and replacing that GitHub variable by the envvar in the relevant `run` command. That is:

```diff
+ env:
+   ISSUE_BODY: ${{ github.event.issue.body}}
+
 jobs:
   foo:
     steps:
-      - run: echo "${{ github.event.issue.body}}"
+      - run: echo "$ISSUE_BODY"
```

Or, on a real example:

```shell
go run main.go \
  --repo pnacht/cronk \
  --commit 6619fad1e79493034363b8865ab5dcbf5442a76c \
  --probes hasDangerousWorkflowScriptInjection \
  --format probe | jq .
```

```json
{
  "...":,
  "findings": [
    {
      "...": "...",
      "remediation": {
        "patch": "<pretty-printed below>",
        "...": "..."
      },
      "probe": "hasDangerousWorkflowScriptInjection",
    }
  ]
}
```

Where the patch is:

```yaml
diff --git a/.github/workflows/awesome-action.yml b/.github/workflows/awesome-action.yml
index 5eeb62ef6d3e94fc63676e9f9557a9389d05a99c..d234fa3415691e8f3c4f40183770fcea755f8d2c 100644
--- a/.github/workflows/awesome-action.yml
+++ b/.github/workflows/awesome-action.yml
@@ -4,6 +4,9 @@   push:
     branches: ["main"]
   pull_request:
 
+env:
+  PR_BODY: ${{ github.event.pull_request.body }}
+
 jobs:
   this_is_safe:
     runs-on: ubuntu-latest
@@ -14,7 +17,7 @@         uses: actions/checkout@v3
 
       - name: "Print PR title"
         if: github.event_name == 'pull_request'
-        run: echo "${{ github.event.pull_request.body }}"
+        run: echo "$PR_BODY"
 
       - name: "Do something awesome"
         uses: super-safe/nothing-to-see-here-action@v1.2.3

```

The patched workflow is then validated by parsing it with `actionlint`. As long as the patch added no new parsing errors, it  is accepted.


#### Which issue(s) this PR fixes

Fixes #3950 

#### Special notes for your reviewer

- This feature requires access to the original workflow files, so we must know the path to the tempdir where the downloaded repository is stored. This is done in `pkg/scorecard_result.populateRawResults`.
- This feature requires a significant amount of "custom parsing". We can't simply patch the workflow struct created by `actionlint.Parse()` because it loses style information (i.e. whitespace, order of elements, etc). We must therefore parse the workflow ourselves to ensure we make minimal patches that follow the original file's style as best we can. However, the logic does use the `actionlint.Workflow` when possible (which is only to read any existing environment variables).
- Regarding the global environment:
  - If the workflow already contains a global environment, it is used.
    - The new envvar adopts the same indentation as the existing envvars.
    - If an existing global envvar already covers the dangerous variable, we use it instead of creating a new envvar with the same value
    - If an existing global envvar has the same name as the one we'd create, but a different value, we append an arbitrary hard-coded string to avoid conflicts. Envvars at lower scopes (job- and step-level) are not considered, which may lead to (likely very rare) conflicts.
  - If a new `env` must be created:
    - it is created right above the `jobs:` label
    - In an attempt to keep the workflow's style:
      - the indentation step used for the new envvar will be copied from the indentation step used for the individual job labels.
      - blanks lines will be inserted between the envvar definition and the `jobs:` label. The number of lines matches the number between the `jobs:` label and the end of the preceding block in the original workflow.
- In case of errors at any point in the process, the patch is simply left blank, without interrupting Scorecard's ordinary flow. I was unsure how to log these errors (is it just a matter of creating a new logger with `sclog.NewLogger(WarnLevel)`?).

Known limitations:
- As mentioned above, the feature does not currently detect potential name collisions between the new environment variable (declared globally) and environment variables declared at a lower scope (job- or step-level).
  This would require a full parsing of the entire workflow to understand precisely which step, in which job, the finding is flagging, which environment variables exist at that step, etc. Given that such name collisions seem exceedingly rare, the current "basic" implementation seems sufficient, in my opinion.
- There are situations where the proper use of the envvar isn't `$FOO`, but `env.foo` (i.e. in a more complex GitHub variable expansion) or `process.env.foo` (i.e. when using `actions/github-script`). The current implementation does not handle these situations properly, and always uses `$FOO`.

Open questions:

- Should this feature be added to the Scorecard documentation? If so, where? `checks.yml/md`?
- The logic to generate a patch diff is pretty generic and could easily be used in in other probes' Remediation.patch implementations. However, I'm unsure where the best place for such features would be. Create a new `remediation/patch.go`?

#### Does this PR introduce a user-facing change?

```release-note
When detecting a potential script injection in a GitHub workflow, Scorecard now adds a machine-readable patch to fix the vulnerability. This patch can be applied to your project using `git apply` or `patch -p1` from the repository's root.
```

Thanks to @joycebrum and @diogoteles08 who helped come up with the tests and the logic to integrate with `hasDangerousWorkflowScriptInjection.Run`.
